experimental/token/raw.go

```diff
@@ -99,18 +104,41 @@ type nat struct {
 	// than the start, it makes adding tokens one by one to the stream
 	// easier, because once the token is pushed, its start and end are
 	// set correctly, and don't depend on the next token being pushed.
-	end           uint32
-	kindAndOffset int32
+	end uint32
+
+	// This contains compressed metadata about a token in the following format:
+	//
+	// 1. Bits [0:3] is the Kind.
+	// 2. Bit [3] is whether this is a non-leaf.
+	//    a. If it is a non-leaf, bits [4:6] determine the Keyword value if
+	//	     Kind is Punct, and bits [6:32] are a signed offset to the matching
+	//       open/close.
+	//    b. If it is a leaf, bits [4:12] are a Keyword value.
+	metadata int32
```

Comment:
This seems to give up quite a few bits for the offset, which is rather unfortunate. It also doesn't seem necessary: you could do this another way, too:
1. If bit 1 (LSB) is 1, then this is a tree (and kind implicitly is Punct).
    * Bits 2-3 indicate the keyword value
2. If bit 1 is 0, then bits 2-4 are the kind. If kind is Ident, bits 5-13 are the keyword value.

Reply:
Well... String tokens can also be fused, which is how we handle concatenation. I don't actually think we give up too many bits. As mentioned, there need to be 32 million tokens between two braces for this to be a problem. I think we should consider burning that bridge when we get there.